### PR TITLE
Stats Plugin Test

### PR DESCRIPTION
**Pull request by gargnupur**
Stats Plugin Test
TODO: Need to do cleanup of testing prometheus metrics

**What this PR does / why we need it**:
Stats Plugin Test

### CODE REVIEW

File: test/envoye2e/env/setup.go [hunk 1]
```diff
@@ -370,6 +381,23 @@ func (s *TestSetup) WaitServerEnvoyReady() {
...
+		aStatsValue, ok := actualStatsMap[eStatsName]
+		if !ok && eStatsValue != 0 {
+			return fmt.Errorf("failed to find expected stat %s", eStatsName)
+		}
+		if aStatsValue != eStatsValue {
+			return fmt.Errorf("stats %s does not match. expected vs actual: %d vs %d",
+				eStatsName, eStatsValue, aStatsValue)
+		}
+
+		log.Printf("stat %s is matched. value is %d", eStatsName, eStatsValue)
```
Comment (venilnoronha) [inline test/envoye2e/env/setup.go]: Wouldn't this cause too many log lines?

File: test/envoye2e/env/setup.go [hunk 2]
```diff
@@ -386,43 +414,53 @@ func (s *TestSetup) unmarshalStats(statsJSON string) map[string]int {
...
+}
+
+func (s *TestSetup) verifyStats(expectedStats map[string]int, statsURL string, unmarshal func(a string) map[string]int) {
 	delay := 200 * time.Millisecond
 	total := 3 * time.Second
 
 	var err error
 	for attempt := 0; attempt < int(total/delay); attempt++ {
-		statsURL := fmt.Sprintf("http://localhost:%d/stats?format=json&usedonly", port)
+		//statsURL := fmt.Sprintf("http://localhost:%d/stats?format=json&usedonly", port)
```
Comment (venilnoronha) [inline test/envoye2e/env/setup.go]: Looks like we could drop this line.

File: test/envoye2e/env/setup.go [hunk 3]
```diff
@@ -386,43 +414,53 @@ func (s *TestSetup) unmarshalStats(statsJSON string) map[string]int {
...
-		for eStatsName, eStatsValue := range expectedStats {
-			aStatsValue, ok := actualStatsMap[eStatsName]
-			if !ok && eStatsValue != 0 {
-				return fmt.Errorf("failed to find expected stat %s", eStatsName)
-			}
-			if aStatsValue != eStatsValue {
-				return fmt.Errorf("stats %s does not match. expected vs actual: %d vs %d",
-					eStatsName, eStatsValue, aStatsValue)
-			}
+// UnmarshalPrometheusStats Unmarshals Envoy Prometheus stats from file format into a map, where stats name is
```
Comment (venilnoronha) [inline test/envoye2e/env/setup.go]: s/UnmarshalPrometheusStats/unmarshalPrometheusStats

File: test/envoye2e/env/setup.go (hunk 1 above)
Comment (gargnupur) [reply test/envoye2e/env/setup.go]: Yeah was good for debugging.. deleted.

File: test/envoye2e/env/setup.go (hunk 3 above)
Comment (gargnupur) [reply test/envoye2e/env/setup.go]: Fixed

File: test/envoye2e/env/setup.go (hunk 2 above)
Comment (gargnupur) [reply test/envoye2e/env/setup.go]: Fixed

File: test/envoye2e/env/setup.go [hunk 4]
```diff
@@ -370,6 +381,21 @@ func (s *TestSetup) WaitServerEnvoyReady() {
 	s.WaitEnvoyReady(s.Ports().ServerAdminPort)
 }
 
+func checkStats(expectedStatsMap map[string]int, actualStatsMap map[string]int) error {
+
+	for eStatsName, eStatsValue := range expectedStatsMap {
+		aStatsValue, ok := actualStatsMap[eStatsName]
+		if !ok && eStatsValue != 0 {
```
Comment (bianpengyuan) [inline test/envoye2e/env/setup.go]: Why it is ok to not find the stat in actual map if expected value is 0? For gauge stat, this seems to be wrong.

File: test/envoye2e/env/setup.go [hunk 5]
```diff
@@ -370,6 +381,21 @@ func (s *TestSetup) WaitServerEnvoyReady() {
 	s.WaitEnvoyReady(s.Ports().ServerAdminPort)
 }
 
+func checkStats(expectedStatsMap map[string]int, actualStatsMap map[string]int) error {
+
```
Comment (bianpengyuan) [inline test/envoye2e/env/setup.go]: nit: extra line?

File: test/envoye2e/env/setup.go [hunk 6]
```diff
@@ -386,43 +412,52 @@ func (s *TestSetup) unmarshalStats(statsJSON string) map[string]int {
...
-			if !ok && eStatsValue != 0 {
-				return fmt.Errorf("failed to find expected stat %s", eStatsName)
-			}
-			if aStatsValue != eStatsValue {
-				return fmt.Errorf("stats %s does not match. expected vs actual: %d vs %d",
-					eStatsName, eStatsValue, aStatsValue)
-			}
+// unmarshalPrometheusStats Unmarshals Envoy Prometheus stats from file format into a map, where stats name is
+// key, and stats value is value. Currently, this is good for counters and not much for histograms.
+func (s *TestSetup) unmarshalPrometheusStats(statsStr string) map[string]int {
```
Comment (bianpengyuan) [inline test/envoye2e/env/setup.go]: Is there any prometheus library to do stats parsing? e.g. https://github.com/prometheus/common/blob/master/expfmt/text_parse.go

File: test/envoye2e/stats_plugin/stats_plugin_test.go [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+package client_test
+
+import (
+	"fmt"
+	"testing"
+
+	"istio.io/proxy/test/envoye2e/env"
+)
+
+const outboundStatsFilter = `- name: envoy.filters.http.wasm
```
Comment (mandarjog) [inline test/envoye2e/stats_plugin/stats_plugin_test.go]: Can we change all these from inline string to just yaml files?
They can be templatized if needed.
These are the authoritative config files, which should be as plain yamls.

File: test/envoye2e/env/setup.go (hunk 4 above)
Comment (gargnupur) [reply test/envoye2e/env/setup.go]: yeah removed that logic... it was carried over from existing envoy stats verification function..

File: test/envoye2e/env/setup.go (hunk 5 above)
Comment (gargnupur) [reply test/envoye2e/env/setup.go]: Removed this function.. not needed anymore

File: test/envoye2e/env/setup.go (hunk 6 above)
Comment (gargnupur) [reply test/envoye2e/env/setup.go]: Thanks for the pointer.. using this..

File: test/envoye2e/stats_plugin/stats_plugin_test.go (hunk 7 above)
Comment (gargnupur) [reply test/envoye2e/stats_plugin/stats_plugin_test.go]: These are already templatized in envoy_conf.go and tcp_envoy_conf.go.. 
the only part that is inline are extra filters or custom data...

like it this way as it shows in test file itself what extra config we are testing..

**Comment by gargnupur** [comment]
@mandarjog , @bianpengyuan , @kyessenov : I added dependencies for prometheus client and thus go.sum got modified.. am getting following error in presubmit runs:
"writing go.sum: open /home/prow/go/src/istio.io/proxy/go.sum224290483.tmp: permission denied"

how to fix these?

**Comment by gargnupur** [comment]
@geeknoid : I added dependencies for prometheus client and thus go.sum got modified.. am getting following error in presubmit runs:
"writing go.sum: open /home/prow/go/src/istio.io/proxy/go.sum224290483.tmp: permission denied"

can you please help on how to fix these?

**Comment by kyessenov** [comment]
Maybe merge master and try again? I also changed go.mod but it worked fine for me.

**Comment by gargnupur** [comment]
@kyessenov : sure, will try that

**Comment by gargnupur** [comment]
@kyessenov : can you please look at this again?

File: test/envoye2e/env/setup.go [hunk 8]
```diff
@@ -432,6 +445,69 @@ func (s *TestSetup) VerifyStats(expectedStats map[string]int, port uint16) {
...
+				return fmt.Errorf("failed to find expected stat %s", eStatsName)
+			}
+			var aStatsValue float64
+			switch aStats.GetType() {
+			case dto.MetricType_COUNTER:
+				log.Printf("found counter")
+				if len(aStats.GetMetric()) != 1 {
+					return fmt.Errorf("expected one value for counter")
+				}
+				aStatsValue = (*(aStats.GetMetric())[0]).GetCounter().GetValue()
```
Comment (kyessenov) [inline test/envoye2e/env/setup.go]: You don't need this pointer dance. `aStats.GetMetric()[0].GetCounter()` works.

File: test/envoye2e/env/setup.go [hunk 9]
```diff
@@ -432,6 +445,69 @@ func (s *TestSetup) VerifyStats(expectedStats map[string]int, port uint16) {
...
 
+// VerifyPrometheusStats verifies prometheus stats.
+func (s *TestSetup) VerifyPrometheusStats(expectedStats map[string]int, port uint16) {
+	s.t.Helper()
+
+	check := func(respBody string) error {
+		for eStatsName, eStatsValue := range expectedStats {
+			var parser expfmt.TextParser
+			reader := strings.NewReader(respBody)
+			mapMetric, err := parser.TextToMetricFamilies(reader)
```
Comment (kyessenov) [inline test/envoye2e/env/setup.go]: Move this out of the loop. Parsing is expensive.

File: test/envoye2e/env/setup.go [hunk 10]
```diff
@@ -432,6 +445,69 @@ func (s *TestSetup) VerifyStats(expectedStats map[string]int, port uint16) {
...
+				if len(aStats.GetMetric()) != 1 {
+					return fmt.Errorf("expected one value for counter")
+				}
+				aStatsValue = (*(aStats.GetMetric())[0]).GetCounter().GetValue()
+				break
+			case dto.MetricType_GAUGE:
+				if len(aStats.GetMetric()) != 1 {
+					return fmt.Errorf("expected one value for gauge")
+				}
+				aStatsValue = (*(aStats.GetMetric())[0]).GetGauge().GetValue()
```
Comment (kyessenov) [inline test/envoye2e/env/setup.go]: Same here. No need to do pointer dance.

File: test/envoye2e/env/setup.go (hunk 9 above)
Comment (gargnupur) [reply test/envoye2e/env/setup.go]: Done.. Thanks!

File: test/envoye2e/env/setup.go (hunk 8 above)
Comment (gargnupur) [reply test/envoye2e/env/setup.go]: Thanks..Fixed

File: test/envoye2e/env/setup.go (hunk 10 above)
Comment (gargnupur) [reply test/envoye2e/env/setup.go]: Thanks.. Fixed

File: test/envoye2e/stats_plugin/stats_plugin_test.go [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+	s.SetFiltersBeforeEnvoyRouterInProxyToServer(inboundStatsFilter)
+	s.SetServerNodeMetadata(inboundNodeMetadata)
+	s.SetClientNodeMetadata(outboundNodeMetadata)
+	s.SetExtraConfig(statsConfig)
+	if err := s.SetUpClientServerEnvoy(); err != nil {
+		t.Fatalf("Failed to setup test: %v", err)
+	}
+	defer s.TearDownClientServerEnvoy()
+
+	url := fmt.Sprintf("http://localhost:%d/echo", s.Ports().AppToClientProxyPort)
```
Comment (kyessenov) [inline test/envoye2e/stats_plugin/stats_plugin_test.go]: nit: please use 127.0.0.1 since DNS is flakey

File: test/envoye2e/env/setup.go (hunk 9 above)
Comment (kyessenov) [reply test/envoye2e/env/setup.go]: Wait, this is not done, right? Still parsing inside the loop.

Comment (gargnupur) [reply test/envoye2e/env/setup.go]: looks like none of my changes got pushed...

Comment (gargnupur) [reply test/envoye2e/env/setup.go]: Fixed now

**Pull request merged**